### PR TITLE
feat(code-block): onCopyCallback

### PR DESCRIPTION
**Pull request by zchsh**
🎟️ [Asana Task](https://app.asana.com/0/1201903760348480/1202114097062999/f)
🔍 [Preview Link](https://react-components-git-zscode-block-on-copy-hashicorp.vercel.app/components/codeblock)

---

This PR implements an `onCopyCallback` prop for `code-block`. This callback is called after `code-block`'s Copy button is clicked.

Note that this callback is _not_ called as a direct result of the `onClick` on the button - instead, we wait until we can confirm whether the copying of text was successful.

### PR Checklist 🚀

Items in this checklist may not may not apply to your PR, but please consider each item carefully.

- [x] Add Asana and Preview links above.
- [x] Conduct thorough self-review.
- [x] Add or update tests as appropriate.
- [x] Conduct reasonable cross browser testing for both compatibility and responsive behavior (We have a [Sauce Labs](https://app.saucelabs.com/) account for this, if you don't have access, just ask!).
- [x] Conduct reasonable accessibility review (use the [WAS](https://accessible.org/Web-Accessibility-Standards-WAS-2.pdf) as a guide or an [axe browser plugin](https://www.deque.com/axe/) until we establish more formal checks).
- [x] Identify (in the description above) and document (add Asana tasks on [this board](https://app.asana.com/0/1100423001970639/list)) any technical debt that you're aware of, but are not addressing as part of this PR.

### CODE REVIEW

File: packages/code-block/partials/clipboard-button/index.tsx [hunk 1]
```diff
@@ -45,6 +50,10 @@ function ClipboardButton({ className, getText }: ClipboardButtonProps) {
   // reset to the default appearance so that it's clear
   // the "Copy" button can be used again
   useEffect(() => {
```
Comment (zchsh) [inline packages/code-block/partials/clipboard-button/index.tsx:52]: We wait till the `copiedState` has changed before calling the callback.

This way, the callback has richer information, as it provides a `copiedState`, rather than only reporting that the copy button was clicked (with an unknown outcome) as we'd get if we hooked into the copy button's `onClick`.

File: packages/code-block/index.test.js [hunk 2]
```diff
@@ -76,6 +77,7 @@ it('should have a data-heap-track="code-block-clipboard-icon" attribute on the C
   )
   const copyButton = screen.getByText('Copy')
   expect(copyButton.tagName).toBe('BUTTON')
+  expect(copyButton.getAttribute('data-heap-track')).toBe(heapAttributes.copy)
```
Comment (zchsh) [inline packages/code-block/index.test.js:80]: This and the `heapAttributes` import are an unrelated fix for a slip I noticed when writing new tests. Previously we weren't actually testing for the `data-heap-track` attribute 😅

**Comment by zchsh** [comment]
@BrandonRomano This change is mainly targeted at enabling https://github.com/hashicorp/dev-portal/pull/443. Let me know what you think 🙇‍♂️

**Review by BrandonRomano — APPROVED** [review]
Nice improvement here ✨ 

Was there an observed problem with the simple click handling, or just hardening to ensure it was actually copied?
